src/App.js

```diff
@@ -22,109 +22,98 @@ mapboxgl.accessToken = config.accessToken;
 export default function App() {
     const mapContainer = useRef(null);
     const map = useRef(null);
-    const scroller = scrollama();
-    const data = createLine(rawData, config.driveSmoothness);
-    const feature = {
+    const line = createLine(rawData, config.driveSmoothness);
+    const empty = {
         type: 'FeatureCollection',
-        features: [
-            {
-                type: 'Feature',
-                geometry: {
-                    type: 'LineString',
-                    coordinates: [],
-                },
-            },
-        ],
+        features: [{ type: 'Feature', geometry: { type: 'LineString', coordinates: [] } }],
     };
-    const [isVisible, setVisibility] = useState(false); 
+    const [isVisible, setVisibility] = useState(false);
 
     const handleChange = () => {
         setVisibility(!isVisible);
     }
 
+    // Init scrollama, destroying the previous instance to prevent leaks
+    let scroller = null;
+    const resetScrollama = () => {
+        if (!map.current) return
+        if (scroller) scroller.destroy()
+        scroller = scrollama()
+            .setup({ step: '.step', offset: 0.5, progress: true })
+            .onStepEnter(response => actions.enter(response, config, map.current))
+            .onStepProgress(response => actions.progress(response, config, map.current, line))
+            .onStepExit(response => actions.exit(response));
+    };
+
+    // Since the map emits the `load` event only after fetching tiles,
+    // we need to init scrollama and emit progress events earlier
+    // to make sure the map doesn't load tiles we aren't interested in.
+    // Interval is a hack, there's probably a better way.
+    // We will remove it once the map finished loading.
+    let scrollReset = setInterval(resetScrollama, 200);
+    // Reinit on react renders
+    useEffect(resetScrollama);
+
     useEffect(() => {
+        window.addEventListener('resize', resetScrollama);
+        return () => window.removeEventListener('resize', resetScrollama);
+    });
+
+    // Init mapbox-gl
+    // The earlier map gets the final set of parameters,
+    // the less time it will waste on fetching / rendering the wrong thing,
+    // so we do our best to calculate all settings before calling the constructor
+    // and not call any methods later on
+    useEffect(() => {
+        if (!mapContainer.current) return;
         if (map.current) return;
 
         const { center, zoom, bearing, pitch } = config.chapters[0].location;
+        const { style } = config;
```

Comment:
Are we passing a local .json just because it will help us organize adding sources and layers? Or is there another performance benefit to this?

Reply:
Another benefit is that we know that the map is completely set up once `map.current` gets set, without a network request to fetch the style that may take a while. This, in turn, allows us to immediately start updating the route line and move camera around, since we know that all essential layers and datastructures are already there by that time.

Reply:
makes sense!